README.md
This project contains the source code and makefile to build the "PackScan" memory pack scanning tool (packscan). packscan is a command-line utility that performs basic parsing and forensics of data dumped from Super Famicom memory packs. It searches a dump to find content headers, determine which blocks of the pack's flash memory contain data, calculate CRCs on content data, and provide other useful information. 

*** Building the Software ***

Because packscan does not rely on any external dependencies, it should build on any Unix-like platform that understands makefiles and has a C++ compiler. Simply build it by running "make".

*** Using the Software ***

Most users will use packscan to generate a report like this:

$ ./packscan [NAME OF DUMP FILE]

Run packscan with a "-h" for a list of other options:

$ ./packscan -h

*** Credits ***

The Shift-JIS to UTF8 conversion code in packscan is copied from the "apollo-ps3" project: https://github.com/bucanero/apollo-ps3. All other source code in this project is original to packscan.

*** Licenses ***

All source code in this project inherits a GPL v3 license from the apollo-ps3 project.

*** Contact ***

The code in this project was written by Andrew Henderson (hendersa@icculus.org) and is part of the BeagleSatella project. BeagleSatella is an effort to develop a software/hardware platform that can interface with SNES/SFC consoles as a external peripheral. Learn more about BeagleSatella here:

https://www.beaglesatella.org



CHANGELOG
CHANGELOG for packscan:

2021-08-18 (v0.1)

- Initial release

2021-11-21 (v0.2)

- Changed the four-byte "program type" check when parsing potential headers. The check is now 0x00000100 for BS-X bytecode (used only by BS-X), 0x00000200 for SA-1 code (used only by Itoi Shigesato no Bass Tsuri No. 1), and 65C816 code for everything else. Thanks to LuigiBlood for the clarification.
- Fixed option parsing.
- Replaced those '\n' EOLs with std::endl.
- Added check to discard a potential header if it has no blocks allocated.
- Changed the main data buffer from std::string to std::vector<uint8_t>.
- Added "-Wall -Werror -pedantic" to CXXFLAGS in the Makefile and cleaned up evertyhing being reported.



pack.h
/****************************************************************
 * PackScan: An SFC memory pack dump analysis tool.
 *
 * Written by Andrew Henderson (hendersa@icculus.org).
 *
 * This code is open source and licensed under the GPLv3. Please
 * review the LICENSE file for the details if you would like to
 * use this source code in your own projects.
 ***************************************************************/

#ifndef __PACK_H__
#define __PACK_H__

#include <string>
#include <vector>
#include <cstdint>

class Pack {
public:
    Pack(const char *filename);
    ~Pack();
    bool isLoaded(void) { return mIsLoaded; }
    void analyze(void);
    std::string generateReport(const bool color);

private:

    std::vector<uint8_t> mPackData;
    std::string mFilename;
    bool mIsLoaded;

    enum PackSize_t {
       INVALID = 0,
       SIZE_8M = (1024 * 1024),
       SIZE_32M = (4 * 1024 * 1024)
    } mPackSize;

    typedef struct {
        uint32_t address;       /* Address of header in pack */
        uint8_t licensee[2];    /* xFB0-xFB1 */
        uint8_t programType[4]; /* xFB2-xFB5 */
        uint8_t title[17];      /* xFC0-xFCF (plus NUL) */
        uint8_t blockAlloc[4];  /* xFD0-xFD3 */
        uint8_t starts[2];      /* xFD4-xFD5 */
        uint8_t dateMonth;      /* xFD6 */
        uint8_t dateDay;        /* xFD7 */
        uint8_t speedMap;       /* xFD8 */
        uint8_t fileType;       /* xFD9 */
        uint8_t maker;          /* xFDA */
        uint8_t version;        /* xFDB */
        uint16_t invChksum;   /* xFDC-xFDD */
        uint16_t chksum;      /* xFDE-xFDF */
    } Header_t;

    std::vector<Header_t> mBlockHeader;

    bool validHeader(const uint32_t block, const bool LoROM, Pack::Header_t *header);
    uint16_t calcCRC(const Header_t *header);
};

#endif /* __PACK_H__ */



main.cpp
/****************************************************************
 * PackScan: An SFC memory pack dump analysis tool.
 *
 * Written by Andrew Henderson (hendersa@icculus.org).
 *
 * This code is open source and licensed under the GPLv3. Please
 * review the LICENSE file for the details if you would like to
 * use this source code in your own projects.
 ***************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <iostream>
#include "version.h"
#include "pack.h"

static void showVersion(void)
{
    std::cout << std::endl;
    std::cout << "PackScan: An SFC memory pack dump analyzer (v";
    std::cout << VERSION << ")" << std::endl;
    std::cout << "Written by Andrew Henderson (hendersa@icculus.org)";
    std::cout << std::endl;
    std::cout << "Source available: https://www.beaglesatella.org";
    std::cout << std::endl << std::endl;
}

static void showHelp(const char *programName) 
{
    std::cout << "Usage:" << std::endl << std::endl << "  " << programName;
    std::cout << " [options] [Memory pack dump filename]" << std::endl;
    std::cout << std::endl << "Options:" << std::endl;
    std::cout << "  -n   No color codes in report" << std::endl;
    std::cout << "  -v   Display version" << std::endl;
    std::cout << "  -h   Display this help" << std::endl;
}

int main(int argc, char *argv[]) 
{
    Pack *pack = NULL;
    int fileIdx = 0;
    bool useColor = true;
    int opt = 0;

    /* Parse command line options */
    while ((opt = getopt(argc, argv, "nvh")) != -1)
    {
        switch(opt)
        {
            case 'n':
                useColor = false;
                break;

            case 'v':
                showVersion();
                return 0;

            case 'h':
                showVersion();
                showHelp(argv[0]);
                return 0;

            default:
                break;		
        }
    } /* End while */

    /* Parse memory pack dump filename */
    if ( optind == (argc - 1) )
    {
        fileIdx = optind;
    }
    else
    {
        /* Filename wasn't specified */
        showVersion();
	std::cout << "No memory pack file specified." << std::endl << std::endl;
        showHelp(argv[0]);
        return 0;
    }

    /* Load the pack data */
    pack = new Pack(argv[fileIdx]);
    if (!pack->isLoaded())
    {
        delete pack;
	return 0;
    }

    /* Analyze the pack and generate a report */
    pack->analyze();
    showVersion();
    std::cout << pack->generateReport(useColor);

    /* Delete the pack data and exit */
    delete pack;
    return 0;
}



shiftjis_conv.cpp

/* The code in this file was originally part of the "apollo-ps3" 
 * project, licensed under GPLv3. You can find the original code 
 * for these functions in the "psv_resign.c" file of the project:
 *
 * https://github.com/bucanero/apollo-ps3/blob/master/source/psv_resign.c
 */

#include <string.h>
#include <stdio.h>
#include <stdlib.h>

#include <stdint.h>
#include "shiftjis_table.h"

const char SJIS_REPLACEMENT_TABLE[] = 
    " ,.,..:;?!\"*'`*^"
    "-_????????*---/\\"
    "~||--''\"\"()()[]{"
    "}<><>[][][]+-+X?"
    "-==<><>????*'\"CY"
    "$c&%#&*@S*******"
    "*******T><^_'='";


//Convert Shift-JIS characters to ASCII equivalent
void sjis2ascii(char* bData)
{
	uint16_t ch;
	int i, j = 0;
	int len = strlen(bData);
	
	for (i = 0; i < len; i += 2)
	{
		ch = (bData[i]<<8) | bData[i+1];

		// 'A' .. 'Z'
		// '0' .. '9'
		if ((ch >= 0x8260 && ch <= 0x8279) || (ch >= 0x824F && ch <= 0x8258))
		{
			bData[j++] = (ch & 0xFF) - 0x1F;
			continue;
		}

		// 'a' .. 'z'
		if (ch >= 0x8281 && ch <= 0x829A)
		{
			bData[j++] = (ch & 0xFF) - 0x20;
			continue;
		}

		if (ch >= 0x8140 && ch <= 0x81AC)
		{
			bData[j++] = SJIS_REPLACEMENT_TABLE[(ch & 0xFF) - 0x40];
			continue;
		}

		if (ch == 0x0000)
		{
			//End of the string
			bData[j] = 0;
			return;
		}

		// Character not found
		bData[j++] = bData[i];
		bData[j++] = bData[i+1];
	}

	bData[j] = 0;
	return;
}

// PSV files (PS1/PS2) savegame titles are stored in Shift-JIS
char* sjis2utf8(char* input)
{
    // Simplify the input and decode standard ASCII characters
    sjis2ascii(input);

    size_t len = strlen(input);
    char* output = (char *)malloc(3 * len); //ShiftJis won't give 4byte UTF8, so max. 3 byte per input char are needed
    size_t indexInput = 0, indexOutput = 0;

    while(indexInput < len)
    {
        char arraySection = ((uint8_t)input[indexInput]) >> 4;

        size_t arrayOffset;
        if(arraySection == 0x8) arrayOffset = 0x100; //these are two-byte shiftjis
        else if(arraySection == 0x9) arrayOffset = 0x1100;
        else if(arraySection == 0xE) arrayOffset = 0x2100;
        else arrayOffset = 0; //this is one byte shiftjis

        //determining real array offset
        if(arrayOffset)
        {
            arrayOffset += (((uint8_t)input[indexInput]) & 0xf) << 8;
            indexInput++;
            if(indexInput >= len) break;
        }
        arrayOffset += (uint8_t)input[indexInput++];
        arrayOffset <<= 1;

        //unicode number is...
        uint16_t unicodeValue = (shiftJIS_convTable[arrayOffset] << 8) | shiftJIS_convTable[arrayOffset + 1];

        //converting to UTF8
        if(unicodeValue < 0x80)
        {
            output[indexOutput++] = unicodeValue;
        }
        else if(unicodeValue < 0x800)
        {
            output[indexOutput++] = 0xC0 | (unicodeValue >> 6);
            output[indexOutput++] = 0x80 | (unicodeValue & 0x3f);
        }
        else
        {
            output[indexOutput++] = 0xE0 | (unicodeValue >> 12);
            output[indexOutput++] = 0x80 | ((unicodeValue & 0xfff) >> 6);
            output[indexOutput++] = 0x80 | (unicodeValue & 0x3f);
        }
    }

	//remove the unnecessary bytes
    output[indexOutput] = 0;
    return output;
}



pack.cpp
/****************************************************************
 * PackScan: An SFC memory pack dump analysis tool.
 *
 * Written by Andrew Henderson (hendersa@icculus.org).
 *
 * This code is open source and licensed under the GPLv3. Please
 * review the LICENSE file for the details if you would like to
 * use this source code in your own projects.
 ***************************************************************/

#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fstream>
#include <streambuf>
#include <sstream>
#include <iostream>
#include <iomanip>
#include "pack.h"
#include "shiftjis_conv.h"

Pack::Pack(const char *filename) : 
    mIsLoaded(false), mPackSize(INVALID) 
{
    struct stat fileStat;
    int retVal = 0;

    mPackData.empty();
    retVal = stat(filename, &fileStat);

    /* Can we stat() the file? */
    if (retVal == -1) {
        switch(errno) {
            case EACCES:
                std::cout << "Unable to access file '" << filename;
                std::cout << "': Access denied" << std::endl;
	        return;

	    case ENOENT:
                std::cout << "Unable to access file '" << filename;
                std::cout << "': Path doesn't exist" << std::endl;
	        return;

            default:
                std::cout << "Unable to access file '" << filename;
                std::cout << "': Error opening file" << std::endl;
	        return;
        } /* End case */
    } /* End if */

    /* Is the file an actual file? */
    if ((fileStat.st_mode & S_IFMT) != S_IFREG) 
    {
        std::cout << "Unable to access file '" << filename;
        std::cout << "': Not a file" << std::endl;
	return;
    }

    /* Is the file the right size for a valid dump? */
    switch(fileStat.st_size) 
    {
        case SIZE_8M:
        case SIZE_32M:
            mPackSize = static_cast<Pack::PackSize_t>(fileStat.st_size);
	    break;

        default:
            std::cout <<"Dump '" << filename << "' is invalid size (";
            std::cout << fileStat.st_size << " bytes)" << std::endl;
	    return;
    }

    /* Load file data into mPackData */
    mFilename = std::string(filename);
    std::ifstream str(filename);
    mPackData.reserve(mPackSize);
    mPackData.assign((std::istreambuf_iterator<char>(str)), 
        std::istreambuf_iterator<char>());

    /* Done! */
    mIsLoaded = true;
}

Pack::~Pack()
{
    if (mIsLoaded) mPackData.empty();
}

void Pack::analyze(void) 
{
    uint32_t i = 0;
    uint32_t totalBlocks = 0;
    Header_t header;

    if (!mIsLoaded) {
        mBlockHeader.empty();	    
        return;
    }

    switch (mPackSize) {
        case SIZE_8M:
            totalBlocks = 8;
            break;

        case SIZE_32M:
            totalBlocks = 32;
            break;

        default:
            totalBlocks = 0;
    } /* End switch */

    for (i=0; i < totalBlocks; i++) 
    {
        /* Check block for a header */
        if ( validHeader(i, true, &header) )
            mBlockHeader.push_back(header);
        else if ( validHeader(i, false, &header) )
            mBlockHeader.push_back(header);
	
    } /* End for */
}

bool Pack::validHeader(const uint32_t block, const bool LoROM, Pack::Header_t *header) 
{
    uint32_t offset = 0;
    uint32_t i = 0;

    memset(header, 0, sizeof(Pack::Header_t));

    /* Check for a header */
    if (LoROM) offset = (block * 0x10000) + 0x7FB0;
    else offset = (block * 0x10000) + 0xFFB0;

    /* Copy header address */
    header->address = offset;

    /* Copy licensee */
    for (i=0; i < 2; i++)
        header->licensee[i] = mPackData[offset + i + 0x00];

    /* Copy program type */
    for (i=0; i < 4; i++)
        header->programType[i] = mPackData[offset + i + 0x02];

    /* Copy title */
    for (i=0; i < 16; i++)
        header->title[i] = mPackData[offset + i + 0x10];
    header->title[16] = '\0';

    /* Copy block allocation flags */
    for (i=0; i < 4; i++)
        header->blockAlloc[i] = mPackData[offset + i + 0x20];
  
    /* Check for valid block allocation */
    if (mPackData.size() != SIZE_8M)
    {
        if ( (header->blockAlloc[3] != 0) ||
            (header->blockAlloc[2] != 0) ||
            (header->blockAlloc[1] != 0) )
            /* Allocating blocks beyond an 8M datapack */
            return false;
    } else {
        if (header->blockAlloc[0] == 0)
            return false;
    }

    /* Copy limited starts */
    for (i=0; i < 2; i++)
        header->starts[i] = mPackData[offset + i + 0x24];

    /* Copy month/day */
    header->dateMonth = mPackData[offset + 0x26];
    header->dateDay = mPackData[offset + 0x27];

    /* Heuristic: If the date bytes are 0xFFFF, this header is invalid */
    if ((header->dateMonth == 0xFF) && (header->dateDay == 0xFF))
        return false;

    /* Copy map mode */
    header->speedMap = mPackData[offset + 0x28];

    /* Copy file type */
    header->fileType = mPackData[offset + 0x29];

    /* Copy the fixed maker field (modified by BS-X on download) */
    header->maker = mPackData[offset + 0x2A];

    /* Check for valid version number */
    switch(mPackData[offset + 0x2A])
    {
        case 0x33: /* BS-X validated (BS-X changed this to 0x33) */
        case 0xFF: /* Download data, not yet BS-X validated */ 
        case 0x00: /* Deleted data */
            break;

        default:   /* Invalid, can't be a header */
            return false;
    }

    /* Copy version */
    header->version = mPackData[offset + 0x2B];

    /* Copy inverse checksum */
    header->invChksum = (uint8_t)mPackData[offset + 0x2C];
    header->invChksum += ((uint8_t)mPackData[offset + 0x2D]) << 8;

    /* Copy checksum */
    header->chksum = (uint8_t)mPackData[offset + 0x2E];
    header->chksum += ((uint8_t)mPackData[offset + 0x2F]) << 8;
    return true;    
}

std::string Pack::generateReport(const bool color) 
{
    std::stringstream report;
    uint32_t i = 0, x = 0;
    uint32_t temp = 0;
    uint16_t tempCRC = 0;

    std::string colorReset = "";
    std::string colorLabel = "";
    std::string colorGood = "";
    std::string colorBad = "";

    /* If no memory pack is loaded, we're done */
    if (!mIsLoaded) 
    {
        report << "No memory pack is loaded, exiting..." << std::endl;
	return report.str();
    }

    /* If the report is in color, set up the terminal color codes */
    if (color)
    {
        colorReset = "\u001b[0m";
        colorLabel = "\u001b[33m"; /* Yellow */
        colorGood = "\u001b[32m";  /* Green */
	colorBad = "\u001b[31m";   /* Red */
    }

    report << colorLabel << "MEMORY PACK FILENAME: " << colorReset;
    report << mFilename << std::endl;
    report << colorLabel << "MEMORY PACK SIZE:     " << colorReset;
    report << mPackSize << " bytes" << std::endl;

    for(i=0; i < mBlockHeader.size(); i++) 
    {
        report << std::endl;
        report << std::dec << colorLabel << "HEADER #" << (i + 1);
        report << " (offset 0x" << std::uppercase << std::setfill('0');
        report << std::setw(5) << std::hex << (int)mBlockHeader[i].address;
        report << "):" << std::endl << std::dec << std::setw(1);

        report << "    TITLE:" << colorReset << "                [";
        report << sjis2utf8((char *)mBlockHeader[i].title) << "]";
        report << std::endl;

	report << colorLabel << "    DATE:" << colorReset;
	report << "                 ";
        report << (mBlockHeader[i].dateMonth >> 4) << "/";
        report << (mBlockHeader[i].dateDay >> 3) << std::endl;

	report << colorLabel << "    REPORTED CRC/INVERSE:" << colorReset;
        report << " 0x" << std::uppercase << std::setfill('0') << std::setw(4);
        report << std::hex << (int)mBlockHeader[i].chksum;
        report << "/0x";
        report << std::uppercase << std::setfill('0') << std::setw(4);
        report << std::hex << (int)mBlockHeader[i].invChksum;
	if ((mBlockHeader[i].chksum + mBlockHeader[i].invChksum) == 0xFFFF) 
            report << colorGood << " [LOOKS OK!]";
	else
            report << colorBad << " [LOOKS BAD]";
        report << std::endl;

	report << colorLabel << "    CALCULATED CRC:" << colorReset;
	report << "       0x";
        tempCRC = calcCRC(&(mBlockHeader[i]));
        report << std::uppercase << std::setfill('0') << std::setw(4);
        report << std::hex << tempCRC << std::dec;
        if (tempCRC == mBlockHeader[i].chksum)
            report << colorGood << " [MATCHES REPORTED]";
        else
            report << colorBad << " [DOES NOT MATCH REPORTED]";
	report << std::endl;

        report << colorLabel << "    BS-X MENU VISIBILITY: " << colorReset;
	if ( ((mBlockHeader[i].chksum == 0) && 
            (mBlockHeader[i].invChksum == 0)) || 
            (mBlockHeader[i].maker != 0x33) ||
            (mBlockHeader[i].starts[1] == 0x80) )
            report << "No" << colorBad << " [NOT SHOWN IN MENU]";
	else
            report << "Yes" << colorGood << " [SHOWS IN MENU]";
        report << std::endl;

	report << std::dec << colorLabel << "    PROGRAM TYPE:";
        report << colorReset << "         ";
        if ( !mBlockHeader[i].programType[0] &&
            !mBlockHeader[i].programType[1] &&
            !mBlockHeader[i].programType[3] ) {

            switch (mBlockHeader[i].programType[2]) {
                case 0x01:
                    report << "BS-X bytecode";
                    break;

                case 0x02:
                    report << "SA-1 code";
                    break;

                default:
                    report << "65C816 code";
                    break;
            } /* End switch */

        } else
            report << "65C816 code";
        report << std::endl;

	report << colorLabel << "    BOOTS REMAINING:";
        report << colorReset << "      ";
        if ( (mBlockHeader[i].starts[1]) & 0x80 )
        {
            report << ( (mBlockHeader[i].starts[1] >> 2) & 0x1F );
        }
        else
            report << "Unlimited";
        report << std::endl;

        report << colorLabel << "    ROM MAPPING/SPEED:    " << colorReset;
        if ( mBlockHeader[i].speedMap & 1 )
            report << "HiROM,";
        else
            report << "LoROM,";

	if ( (mBlockHeader[i].speedMap >> 4) > 2 )
            report << "FastROM (120 ns)";
        else
            report << "SlowROM (200 ns)";
        report << std::endl;

	report << colorLabel << "    EXECUTION:            " << colorReset;
	if (mBlockHeader[i].fileType & 0x20)
            report << "PSRAM,";
        else
            report << "FLASH,";
	if (mBlockHeader[i].fileType & 0x10)
            report << "Soundlink MUTED,";
	else
            report << "Soundlink UNMUTED,";
        if (mBlockHeader[i].fileType & 0x80)
            report << "NO St. GIGA intro";
        else
            report << "St. GIGA intro";
        report << std::endl;

        report << colorLabel << "    BLOCK ALLOCATION:" << colorReset;
	report << "     [";
        temp = (mBlockHeader[i].blockAlloc[3] << 24);
        temp |= (mBlockHeader[i].blockAlloc[2] << 16);
	temp |= (mBlockHeader[i].blockAlloc[1] << 8);
	temp |= (mBlockHeader[i].blockAlloc[0] << 0);
	for (x=0; x < (uint32_t)(mPackSize / (128 * 1024)); x++)
        {
            if ( (temp >> x) & 0x1 )
                report << "X";
            else
                report << ".";
        } /* End for */
        
       report << "]" << std::endl;
    }

    return report.str();
}

uint16_t Pack::calcCRC(const Pack::Header_t *header)
{
    uint16_t crc = 0;
    uint32_t i = 0;
    uint32_t x = 0;
    uint32_t bitmask = 0;
    uint32_t totalBlocks = mPackSize / (128 * 1024);
    
    bitmask =  (header->blockAlloc[3] << 24);
    bitmask |= (header->blockAlloc[2] << 16);
    bitmask |= (header->blockAlloc[1] << 8);
    bitmask |= (header->blockAlloc[0] << 0);

    for (x = 0; x < totalBlocks; x++)
    {
        /* Is the current block in use by this header? */
        if ( (bitmask >> x) & 0x1 )
        {
            /* Does the current block contain the header? */
            if ( ((x * 0x20000) < header->address) &&
                 (((x+1) * 0x20000) > header->address) )
            {
                /* Calculate CRC before header location */
                for (i = (x * 0x20000); i < header->address; i++)
                    crc += (uint8_t)mPackData[i];

                /*  Calculate CRC after header location */
                for (i = (header->address + 0x30); i < ((x+1) * 0x20000); i++ )
                    crc += (uint8_t)mPackData[i];
            }
            else
            {
                /* Calculate CRC of entire block */
                for (i = (x * 0x20000); i < ((x+1) * 0x20000); i++)
                    crc += (uint8_t)mPackData[i];
            }
        } /* End if */
    } /* End for */

    /* Done! */
    return crc;
}

